# Lab 02 - Hooke's Law

# Housekeeping Tasks before you get started:
## Ensure Widget State Saving is enabled:
1. Please click on Settings and ensure there is a check-mark in front of: Save Widget State Automatically. In Lab 00 you were asked to turn this setting on, but 
if there is no check-mark, then click on the setting to turn it on. (The setting is a toggle, so if you click on it when it is on, it will turn it off. Please ensure it is turned on!)

2. Reload your browser tab (eg CTRL-R, or click on the URL and hit enter).


In [121]:
# These should always be at the top (and don't forget to execute them)
%reset -f
import numpy as np
import data_entry2

Edit this cell and provide the information requested:

- Your name: Mary Hosseini
- Your student number: 41004235
- Partner's name: Musa Sayeed


## Part A - Introduction

***TA grading feedback here:***
- The uncertainty propagation in the code is wrong. Here is the right one for your ref:

u_delta_x = np.sqrt(u_x1**2 + u_x2**2)

u_delta_m = np.sqrt(u_m1**2 + u_m2**2)

u_delta_f = g * u_delta_m

u_k = k * np.sqrt((u_delta_f/ delta_f)**2 + (u_delta_x/ delta_x)**2)

In the future I will make sure to double check my formulas to ensure that I wont make any miscalculations

## Part B - Uncertainty Discussions

***Include notes on class discussion here:***
- The third row of the table had the same um2 and um1 which is very unlikely because in um2 the spring is being compressed
- From the prelab we know $A \pm u_A$
- New formula: $\frac{u_A}{A}$
- How to reduce uncertainty in $\frac{u_k}{k} = \sqrt{(\frac{u_x}{x})^2 + (\frac{u_m}{m})^2}$? do a better job of compressing the spring so that the values are more consistent/stable. 
- Uncertainty decimal place

***Make sure to include a formula for the new tool and a description of how to use it.*** 


- From the prelab we know $A \pm u_A$
- New formula: $\frac{u_A}{A}$

## Part C - First round of measurements

***Include your initial measurement plan in the Markdown cell below, then follow the instructions on setting up your data spreadsheet.***

First we measured the height of the spring. Since the spring was slanted on top, we took the average of the slant by measuring the highest, middle and lowest point of the slant. Then we took the average of our measurements (12.9, 13 and 13.1) and got 13cm as the average total height of the spring.

Then we placed the spring on the scale to get the uncompressed mass of 0.078kg. 

After which I compressed the spring while my partner measured the height of the spring. We recorded the following measurements in the table. For when we were compressing the spring, we took the average of the maximum and minimum number we measured while keeping the compression relatively steady. 

**Side Note:** When measuring the height of the spring, it is important to mention that we measured from the table and subtract the height of the scale from the measurement that we got. So the finalized measurements that we reported is the result after we subtract the height of the scale. 

Uncomment the code below (i.e., remove the '#') to set up your initial spreadsheet 'lab02_data_round1.csv'. Any edits you make will automatically be saved. When you are finished inputting your data, press 'Generate Vectors' to transform each column into a Python vector. For instance, if you title a column 'm1', then a Python vector called 'm1Vec' will be generated. The idea is to use these generated vectors to perform the necessary calculations to obtain $k$ and $u\_k$ for small and large compressions.

In [122]:
 de_round1 = data_entry2.sheet("lab02_data_round1")

Sheet name: lab02_data_round1.csv


In [123]:
# Code Cell to calculate K and uK value
g = 9.801 # acceleration due to gravity

delta_m = m2Vec - m1Vec # change in mass
delta_x = x2Vec - x1Vec # displacement


# 95% CI is 0.01 m for x1
ux1 = 0.01 / 4
# 95% CI is 0.005 m
ux2 = 0.005 / 4
u_delta_x = np.sqrt(ux1Vec**2 + ux2Vec**2) # uncertainty displacement

a = 0.0005
um1 = (a / (np.sqrt(3)))
# 95% CI is 0.005 kg
um2 = 0.005 / 4
u_delta_m = np.sqrt(um1Vec**2 + um2Vec**2) # change in uncertainty mass
# a value for m1 is 0.0005kg

kVec = g * np.abs(delta_m) / np.abs(delta_x)

ukVec = kVec * np.sqrt( (u_delta_m/delta_m)**2 + (u_delta_x/delta_x)**2 )

print("K value (N/m) is: ", kVec)
print("Uncertainty of each k value is: ", ukVec)
print("Relative uncertainty of each k value is: ", ukVec/kVec)

K value (N/m) is:  [113.19464789 112.167      112.303125  ]
Uncertainty of each k value is:  [4.4596959  4.3578935  4.36317371]
Relative uncertainty of each k value is:  [0.03939847 0.03885183 0.03885176]


## Part D - Group Discussion

***Comparisons can be measurement techniques and/or measured/calculated values. Don't forget to provide the names of the group members you compare to.***

## Part 1
### Cristiana's Group:

Cristiana's Group: Compressing the spring as much as possible but not completely and compressing the spring with a notebook to keep the compression more steady/constant. Did not account for the weight of the notebook because the group assumed the weight of their hand was similar to the notebook. We suspect that their K value is somewhat higher than ours due to the weight of the notebook compared the weight of a human hand

k = [131.769, 133.916]

u_k = [2.087, 2.455]

#### 1) Were your values and uncertainties for k different to those of other students? If not, what might have caused these differences?
 
Cristiana's Group First Round values: k = [131.769, 133.916]

Our Group First Round values: k = [113.19464789 112.167      112.303125  ]

Comparing our K values to group 1's, theirs are slightly higher at ~132 N/m and ours at ~113 N/m. However, since our springs are different it doesn't matter too much.

#### 2) Was the relative uncertainty in your k results similar to those of other students? What were other students' measurement strategies?
 
Cristiana's Group First Round values: u_K [0.0158, 0.01833]

Our Group First Round values: u_K = [4.4596959  4.3578935  4.36317371]

Comparing our u_K values to group 1's, ours are double compared to theirs of ~2.2 and ours at ~4.3. This makes sense as they used a notebook which has a different weight compared to a human hand (the notebook was lighter).

#### 3) What are various methods to improve your measurements? (Remember that we want you to improve your results by doing your best possible job of the individual measurements, not by combining your results from multiple sets of measurements.)
 
 Comparing our Relative Uncertainty of K to group 1's, ours are higher compared to theirs, ~0.038 and ~0.016 respectively.

The methods we find that will help us improve our measurements is to find a more stable item/force to compress the spring, that will give us a more a accurate m2 value as well as, make is more easier/accurate for us to measure the x2 using the ruler, since the height will not be fluctuating.

## Part 2

### How we wish to improve our measure approach

- Use a more stable weight/object to compress the spring,
    - This will lead to more accurate measurement of x2, as well as, more stable (less fluctuating) output of the scale for m2.
    - This will lead to a smaller uncertainty value for x2 and m2, which will provide us with a smaller relative uncertainty value due to the equation (a).
- Use a higher weight/object to get a higher delta m and delta x.
    - Since delta m and delta x is in the denominator of the equation (a), this will give us a lower relative uncertainty since the difference of the uncompressed spring vs the compressed spring will be higher.
- Run more tests.
    - By running more tests we are able to achieve an overall average and negate extremes, leading to a better/lower relative uncertainty
- Use a pencil or a straight object to make sure our reading on the ruler is level with the weight/object holding down the spring.
    - Making sure we use a straight object to get the proper measurement on the ruler will get us a lower uncertainty.


Figure. (a)

$$ \frac{u_k}{k} = k\sqrt{\left(\frac{u\_\Delta x}{\Delta x}\right)^2+\left(\frac{u\_\Delta m}{\Delta m}\right)^2}$$

## Part E - Cycles of Improvement

Uncomment the code below to set up a new spreadsheet for your improvement round data. This code will copy your old spreadsheet so that all you have to do is overwrite the data from your initial round with your improvement round data. *IMPORTANT NOTE: do not run the code below until you are finished with your data entry in Part D.* 

In [134]:
de_round2 = data_entry2.sheet_copy("lab02_data_round1", "lab02_data_round2")

Sheet name: lab02_data_round2.csv


In [135]:
# Code Cell to calculate K and uK value
g = 9.801 # acceleration due to gravity

delta_m = m2Vec - m1Vec # change in mass
delta_x = x2Vec - x1Vec # displacement


# 95% CI is 0.005 m for x1
ux1 = 0.005 / 4
# 95% CI is 0.0025 m
ux2 = 0.0025 / 4
u_delta_x = np.sqrt(ux1Vec**2 + ux2Vec**2) # uncertainty displacement

a = 0.0005
um1 = (a / (np.sqrt(3)))
# 95% CI is 0.0025 kg
um2 = 0.0025 / 4
u_delta_m = np.sqrt(um1Vec**2 + um2Vec**2) # change in uncertainty mass
# a value for m1 is 0.0005kg

kVec = g * np.abs(delta_m) / np.abs(delta_x)

ukVec = kVec * np.sqrt( (u_delta_m/delta_m)**2 + (u_delta_x/delta_x)**2 )

print("K value (N/min) is: ", kVec)
print("Uncertainty of each k value is: ", ukVec)
print("Relative uncertainty of each k value is: ", ukVec/kVec)

K value (N/min) is:  [125.07942857 122.34731461 129.86325   ]
Uncertainty of each k value is:  [2.08254754 1.9226815  2.67080581]
Relative uncertainty of each k value is:  [0.0166498  0.01571495 0.02056629]


## Summary of Results

 Using a heavier and more steady object, an ipad instead of a hand, we were able to decrease the uncertainty in our calculations. We used similar measurements for the second part of our lab compared to the first part, the only difference being the ipad placed on a spring instead of the hand. 

#### 1) A brief summary of how your key results and the results of your comparisons evolved throughout the lab.

#### 2)  Report your final spring constant following the significant figure rules.

Our final spring constant is 122.35N/m with an uncertainty 0.016. This spring constant was chosen because it has the lowest relative uncertainty compared to our other values. 

#### 3) 1-2 sentences to describe and justify your major improvements and how they proved to be successful or unsuccessful. It may be helpful to use relative uncertainty to quantify your argument.

Our improvement included using a constant weighted object to use to compress the spring with. We thought that having having an object such as a phone or an ipad would help with having more accurate results compared to compressing with a human hand because it would increase delta m and decrease delta x. We used an ipad instead of a human hand to compress the spring which gave us more accurate results as it decreased the relative uncertainty slightly. Our previous data had a relative uncertainty of [0.03939847 0.03885183 0.03885176]. Whereas our new data had a relative uncertainty of [0.0166498  0.01571495 0.02056629]. As seen from our report, the spring uncertainty decreased throughout the three trials that we did. 

#### 4) 1-2 sentences addressing the experimental goal and describing the conclusions you can make from the results of your comparisons.The lab notebook template already provides some tips on what to include in this section.

The experimental goal of attempting to decrease the relative uncertainty of our measured spring constant was closely achieved by concluding two things from our results. Achieve a higher delta x and delta m, as well as, reducing our uncertainty for x1, x2, and m2 by using techniques to measure each respective variable more effectively, which leads to a lower Confidence Interval.

# Submit

Steps for submission:

1. Click: Run => Run_All_Cells
2. Read through the notebook to ensure all the cells executed correctly and without error.
3. Correct any problems you find.
4. File => Save_and_Export_Notebook_As->HTML
5. Upload the HTML document to the lab submission assignment on Canvas.

In [116]:
# This cell will produce copies of any spreadsheets you have created. Please do not modify.
display_sheets()

Sheet: de_round1   File: lab02_data_round1.csv


x2     m2     x1     m1      ux2     ux1                    um1  \
Units      m     kg      m     kg        m       m                     kg   
0      0.058  0.898  0.129  0.078  0.00125  0.0025  0.0002886751345948129   
1      0.057  0.902  0.129  0.078  0.00125  0.0025  0.0002886751345948129   
2      0.057  0.903  0.129  0.078  0.00125  0.0025  0.0002886751345948129   

           um2  
Units       kg  
0      0.00125  
1      0.00125  
2      0.00125

Sheet: de_round2   File: lab02_data_round2.csv


x2     m2     x1     m1      ux2     ux1                    um1  \
Units      m     kg      m     kg        m       m                     kg   
0      0.045  1.150  0.129  0.078  0.00125  0.0025  0.0002886751345948129   
1      0.040  1.189  0.129  0.078  0.00125  0.0025  0.0002886751345948129   
2      0.061  0.979  0.129  0.078  0.00125  0.0025  0.0002886751345948129   

           um2  
Units       kg  
0      0.00125  
1      0.00125  
2      0.00125